cpp/arcticdb/version/test/test_version_map.cpp

```diff
@@ -575,6 +575,300 @@ TEST(VersionMap, StorageLogging) {
+    ASSERT_FALSE(version_map->has_cached_entry(id, LoadParameter{LoadType::LOAD_FROM_TIME, static_cast<timestamp>(1234)}));
+    ASSERT_FALSE(version_map->has_cached_entry(id, LoadParameter{LoadType::LOAD_UNDELETED}));
+    ASSERT_FALSE(version_map->has_cached_entry(id, LoadParameter{LoadType::LOAD_DOWNTO, static_cast<SignedVersionId>(1)}));
+
+    const auto reloaded = version_map->check_reload(
+            store,
+            id,
+            LoadParameter{LoadType::LOAD_DOWNTO, static_cast<SignedVersionId>(0)},
+            __FUNCTION__);
+
+    ASSERT_EQ(reloaded->get_indexes(false).size(), 2);
+}
+
+TEST(VersionMap, CacheInvalidationAfterLoadFromTimeTombstoned) {
+    using namespace arcticdb;
+    ScopedConfig sc("VersionMap.ReloadInterval", std::numeric_limits<int64_t>::max());
+    auto store = std::make_shared<InMemoryStore>();
+
+    auto version_map = std::make_shared<VersionMap>();
+
+    StreamId id{"test"};
+    auto entry = write_two_versions(store, version_map, id);
+
+    uint64_t ver_to_delete = 1;
+    auto tombstone = version_map->write_tombstone(store, ver_to_delete, id, entry);
+
+    version_map->check_reload(
+            store,
+            id,
+            LoadParameter{LoadType::LOAD_FROM_TIME, static_cast<timestamp>(2)},
+            __FUNCTION__);
+
+    // LOAD_FROM_TIME always goes at least as far back as the latest undeleted version, even if that is earlier than
+    // the specified load_from_time_or_until
+    ASSERT_TRUE(version_map->has_cached_entry(id, LoadParameter{LoadType::LOAD_LATEST_UNDELETED}));
+    ASSERT_TRUE(version_map->has_cached_entry(id, LoadParameter{LoadType::LOAD_LATEST}));
+
+    const auto reloaded = version_map->check_reload(
+            store,
+            id,
+            LoadParameter{LoadType::LOAD_LATEST_UNDELETED},
+            __FUNCTION__);
+
+    ASSERT_EQ(reloaded->get_indexes(false).size(), 1);
+    ASSERT_EQ(reloaded->get_indexes(true).size(), 2);
+}
+
+TEST(VersionMap, CacheInvalidationLoadAllAfterLoadFromTime) {
+    using namespace arcticdb;
+    ScopedConfig sc("VersionMap.ReloadInterval", std::numeric_limits<int64_t>::max());
+    auto store = std::make_shared<InMemoryStore>();
+
+    auto version_map = std::make_shared<VersionMap>();
+
+    StreamId id{"test"};
+    write_two_versions(store, version_map, id);
+
+    auto reloaded = version_map->check_reload(
+            store,
+            id,
+            LoadParameter{LoadType::LOAD_FROM_TIME, static_cast<timestamp>(1)},
+            __FUNCTION__);
+
+    ASSERT_FALSE(version_map->has_cached_entry(id, LoadParameter{LoadType::LOAD_ALL}));
+    ASSERT_EQ(reloaded->get_indexes(false).size(), 1);
+
+    reloaded = version_map->check_reload(
+            store,
+            id,
+            LoadParameter{LoadType::LOAD_ALL},
+            __FUNCTION__);
+
+    ASSERT_EQ(reloaded->get_indexes(false).size(), 2);
+}
+
+TEST(VersionMap, CacheInvalidationAfterLoadUndeleted) {
```

Comment:
Same for `CacheInvalidationAfterLoadAll`

Reply:
It was deliberate but a bit of a lazy way to avoid more detailed assertions, so I've added some writes and indirect tests of the cache invalidation for these